orc8r/cloud/go/services/bootstrapper/servicers/registration/storage.go

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright 2020 The Magma Authors.
+
+This source code is licensed under the BSD-style license found in the
+LICENSE file in the root directory of this source tree.
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package registration
+
+import (
+	"github.com/pkg/errors"
+
+	"magma/orc8r/cloud/go/blobstore"
+	"magma/orc8r/cloud/go/services/bootstrapper"
+	"magma/orc8r/cloud/go/storage"
+	merrors "magma/orc8r/lib/go/errors"
+	"magma/orc8r/lib/go/protos"
+)
+
+const networkWildcard = "*"
+
+type Store interface {
+	SetTokenInfo(tokenInfo protos.TokenInfo) error
+	GetTokenInfoFromLogicalID(networkID string, logicalID string) (*protos.TokenInfo, error)
+	GetTokenInfoFromNonce(nonce string) (*protos.TokenInfo, error)
+}
+
+type blobstoreStore struct {
+	factory blobstore.StoreFactory
+}
+
+func NewBlobstoreStore(factory blobstore.StoreFactory) Store {
+	return &blobstoreStore{factory}
+}
+
+func (b *blobstoreStore) SetTokenInfo(tokenInfo protos.TokenInfo) error {
+	isUnique, err := b.isNonceUnique(tokenInfo.Nonce)
+	if err != nil {
+		return err
+	}
+	if !isUnique {
+		return errors.Errorf("token is not unique")
+	}
+
+	networkID := tokenInfo.GatewayDeviceInfo.NetworkId
+	logicalID := tokenInfo.GatewayDeviceInfo.LogicalId
+
+	store, err := b.factory.StartTransaction(nil)
+	if err != nil {
+		return err
+	}
+	defer store.Rollback()
+
+	logicalIDBlob, err := tokenInfoToBlob(bootstrapper.LogicalIDToTokenInfo, logicalID, tokenInfo)
+	if err != nil {
+		return err
+	}
+	err = store.Write(networkID, blobstore.Blobs{logicalIDBlob})
+	if err != nil {
+		return err
+	}
+
+	nonceBlob, err := tokenInfoToBlob(bootstrapper.NonceTokenToInfoMap, tokenInfo.Nonce, tokenInfo)
+	if err != nil {
+		return err
+	}
+	err = store.Write(networkWildcard, blobstore.Blobs{nonceBlob})
+	if err != nil {
+		return err
+	}
+
+	return store.Commit()
+}
+
+func (b *blobstoreStore) GetTokenInfoFromLogicalID(networkID string, logicalID string) (*protos.TokenInfo, error) {
+	store, err := b.factory.StartTransaction(nil)
+	if err != nil {
+		return nil, err
+	}
+	defer store.Rollback()
+
+	logicalIDTK := storage.TK{
+		Type: string(bootstrapper.LogicalIDToTokenInfo),
+		Key:  logicalID,
+	}
+	logicalIDBlob, err := store.Get(networkID, logicalIDTK)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenInfo, err := tokenInfoFromBlob(logicalIDBlob)
+	if err != nil {
+		return nil, err
+	}
+
+	return &tokenInfo, store.Commit()
```

Comment:
Here and throughout: rather than returning addr here, return a ptr from tokenInfoFromBlob

Reply:
When return ptr vs value?

Reply:
Consistency is more important for our purposes. I'm fine with either, but let's not be converting back and forth internally. Ptr is probably a little easier since protos generally flow better as ptrs, but up to you